pkg/controllers/provisioning/provisioner.go

```diff
@@ -0,0 +1,138 @@
+limitations under the License.
+*/
+
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/controllers/provisioning/scheduling"
+	"github.com/awslabs/karpenter/pkg/utils/functional"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+)
+
+var (
+	MaxBatchDuration = time.Second * 10
+	MinBatchDuration = time.Second * 1
+)
+
+// Provisioner waits for enqueued pods, batches them, creates capacity and binds the pods to the capacity.
+type Provisioner struct {
+	// State
+	*v1alpha5.Provisioner
+	instanceTypes []cloudprovider.InstanceType
+	pods          chan *v1.Pod
+	results       chan error
+	Stop          context.CancelFunc
+
+	// Dependencies
+	scheduler *scheduling.Scheduler
+	launcher  *Launcher
+}
+
+// NewProvisioner creates a new provisioner and starts the provisioning loop
+func NewProvisioner(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, scheduler *scheduling.Scheduler, launcher *Launcher) *Provisioner {
+	provisioner.Spec.Labels = functional.UnionStringMaps(provisioner.Spec.Labels, map[string]string{v1alpha5.ProvisionerNameLabelKey: provisioner.Name})
+	provisioner.Spec.Requirements = provisioner.Spec.Requirements.
+		With(scheduling.GlobalRequirements(instanceTypes)). // TODO(etarn) move GlobalRequirements to this file
+		With(v1alpha5.LabelRequirements(provisioner.Spec.Labels))
+	ctx, cancelFunc := context.WithCancel(ctx)
+	p := &Provisioner{
+		Provisioner:   provisioner,
+		instanceTypes: instanceTypes,
+		pods:          make(chan *v1.Pod),
+		results:       make(chan error),
+		scheduler:     scheduler,
+		launcher:      launcher,
+		Stop:          cancelFunc,
+	}
+	p.Start(ctx)
+	return p
+}
+
+func (p *Provisioner) Start(ctx context.Context) {
+	go func() {
+		logging.FromContext(ctx).Info("Starting provisioner")
+		for {
+			select {
+			case <-ctx.Done():
+				logging.FromContext(ctx).Info("Stopping provisioner")
+				close(p.pods)
+				close(p.results)
+				return
+			default:
+				if err := p.Provision(ctx); err != nil {
+					logging.FromContext(ctx).Errorf("Provisioning failed, %s", err.Error())
+				}
+			}
+		}
+	}()
+}
+
+func (p *Provisioner) Provision(ctx context.Context) (err error) {
+	// Wait for a batch of pods
+	pods := p.Batch(ctx)
+	// Send results
+	defer func() {
+		for i := 0; i < len(pods); i++ {
+			p.results <- err
+		}
+	}()
+	// Separate pods by scheduling constraints
+	schedules, err := p.scheduler.Solve(ctx, p.Provisioner, p.instanceTypes, pods)
+	if err != nil {
+		return fmt.Errorf("solving scheduling constraints, %w", err)
+	}
+	// Launch capacity and bind pods
+	if err := p.launcher.Launch(ctx, schedules, p.instanceTypes); err != nil {
+		return fmt.Errorf("launching capacity, %w", err)
+	}
+	return nil
+}
+
+// Enqueue a pod to the provisioner and block until it's processed
```

Comment:
I'd probably rename this - `Enqueue` kind of a suggests a "fire and forget" type action (add something to a queue), whereas this is actually adding to a "queue" (channel), then waiting for it to be fully processed (i.e., gone from the channel).

Reply:
Agree.